velox/connectors/hive/storage_adapters/s3fs/tests/S3InsertTest.cpp

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/init/Init.h>
+
+#include "gtest/gtest.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.h"
+#include "velox/connectors/hive/storage_adapters/s3fs/tests/MinioServer.h"
+#include "velox/dwio/parquet/reader/ParquetReader.h"
+#include "velox/exec/TableWriter.h"
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/HiveConnectorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::core;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::connector;
+using namespace facebook::velox::connector::hive;
+using namespace facebook::velox::dwio::common;
+using namespace facebook::velox::test;
+using namespace facebook::velox::filesystems;
+
+class S3InsertTest : public testing::Test, public VectorTestBase {
+ public:
+  static constexpr char const* kMinioConnectionString{"127.0.0.1:7000"};
+  static void SetUpTestSuite() {
+    if (minioServer_ == nullptr) {
+      minioServer_ = std::make_shared<MinioServer>(kMinioConnectionString);
+      minioServer_->start();
+    }
+    ioExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(3);
+    filesystems::registerS3FileSystem();
+    auto hiveConnector =
+        connector::getConnectorFactory(
+            connector::hive::HiveConnectorFactory::kHiveConnectorName)
+            ->newConnector(
+                kHiveConnectorId,
+                minioServer_->hiveConfig(),
+                ioExecutor_.get());
+    connector::registerConnector(hiveConnector);
+  }
+
+  static void TearDownTestSuite() {
+    filesystems::finalizeS3FileSystem();
+    if (minioServer_ != nullptr) {
+      minioServer_->stop();
+      minioServer_ = nullptr;
+    }
+  }
+
+  static std::shared_ptr<MinioServer> minioServer_;
+  static std::unique_ptr<folly::IOThreadPoolExecutor> ioExecutor_;
+};
+std::shared_ptr<MinioServer> S3InsertTest::minioServer_ = nullptr;
+std::unique_ptr<folly::IOThreadPoolExecutor> S3InsertTest::ioExecutor_ =
+    nullptr;
+
+PlanNodePtr createInsertPlan(
+    PlanBuilder& inputPlan,
+    const RowTypePtr& outputRowType,
+    const std::string& outputDirectoryPath,
+    const std::vector<std::string>& partitionedBy = {},
+    const std::shared_ptr<HiveBucketProperty>& bucketProperty = {},
+    const connector::hive::LocationHandle::TableType& outputTableType =
+        connector::hive::LocationHandle::TableType::kNew,
+    const CommitStrategy& outputCommitStrategy = CommitStrategy::kNoCommit) {
+  auto insertTableHandle = std::make_shared<core::InsertTableHandle>(
```

Comment:
can't we use PlanBuilder here?

Reply:
We use the `PlanBuilder` for `tableWrite` below.
`InsertTableHandle` is an argument to that. I don't see an easy way to make this part of the PlanBuilder.

Reply:
Never mind, I get what you are are doing here now. :)